### PR TITLE
[Feature] Payment Request

### PR DESCRIPTION
**Pull request by saurabh6790**
frappe/erpnext#4469

### CODE REVIEW

File: erpnext/accounts/doctype/accounts_settings/accounts_settings.json [hunk 1]
```diff
@@ -110,10 +118,60 @@
...
+   "allow_on_submit": 0, 
+   "bold": 0, 
+   "collapsible": 0, 
+   "fieldname": "message", 
+   "fieldtype": "Text Editor", 
+   "hidden": 0, 
+   "ignore_user_permissions": 0, 
+   "in_filter": 0, 
+   "in_list_view": 0, 
+   "label": "Payment Transaction Message",
```
Comment (nabinhait) [inline erpnext/accounts/doctype/accounts_settings/accounts_settings.json]: Payment Request Message?

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+from erpnext.accounts.utils import get_account_currency, get_balance_on
+from itertools import chain
+
+class PaymentRequest(Document):		
+	def validate(self):
+		self.validate_payment_request()	
+		
+	def validate_payment_request(self):
+		if frappe.db.get_value("Payment Request", {"reference_name": self.reference_name, 
+			"name": ("!=", self.name), "status": "Paid", "docstatus": 1}, "name"):
```
Comment (nabinhait) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: why filtered for Paid?

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+		if self.payment_url:
+			frappe.db.set_value(self.doctype, self.name, "status", "Initiated")
+			
+	def set_paid(self):
+		if frappe.session.user == "Guest":
+			frappe.set_user("Administrator")
+			
+		return self.create_journal_voucher_entry()
+	
+	def create_journal_voucher_entry(self):
```
Comment (nabinhait) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: don't use the term journal_voucher, use journal_entry

**Comment by rmehta** [comment]
Couple of more thoughts.

- ability to send a reminder
- save the default subject and message in "Payment Gateway Account" (not sure if you have already done this)

File: erpnext/accounts/doctype/journal_entry/journal_entry.py [hunk 4]
```diff
@@ -534,7 +536,7 @@ def get_default_bank_cash_account(company, voucher_type, mode_of_payment=None):
 		}
 
 @frappe.whitelist()
-def get_payment_entry_against_order(dt, dn):
+def get_payment_entry_against_order(dt, dn, args=None):
 	ref_doc = frappe.get_doc(dt, dn)
```
Comment (rmehta) [inline erpnext/accounts/doctype/journal_entry/journal_entry.py]: change `args` to specific

File: erpnext/accounts/doctype/journal_entry/journal_entry.py [hunk 5]
```diff
@@ -565,11 +570,13 @@ def get_payment_entry_against_order(dt, dn):
...
-		"is_advance": "Yes"
+		"is_advance": "Yes",
+		"bank_account": args["bank_account"] if args else None,
+		"return_obj": args["return_obj"] if args else None
 	})
 
 @frappe.whitelist()
-def get_payment_entry_against_invoice(dt, dn):
+def get_payment_entry_against_invoice(dt, dn, args=None):
 	ref_doc = frappe.get_doc(dt, dn)
```
Comment (rmehta) [inline erpnext/accounts/doctype/journal_entry/journal_entry.py]: specific `amount`, `bank_account`, `journal_entry`

File: erpnext/accounts/doctype/payment_gateway_account/payment_gateway_account.py [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+	def validate(self):
+		self.update_default_payment_gateway()
+		self.set_as_default()
+	
+	def update_default_payment_gateway(self):
+		if self.is_default:
+			frappe.db.sql("""update `tabPayment Gateway Account` set is_default = 0
+				where is_default = 1 """)
+		
+	def set_as_default(self):
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_gateway_account/payment_gateway_account.py]: `set_as_default_if_not_set`

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+
+	def validate_payment_request(self):
+		if frappe.db.get_value("Payment Request", {"reference_name": self.reference_name, 
+			"name": ("!=", self.name), "status": ("not in", ["Initiated", "Paid"]), "docstatus": 1}, "name"):
+			frappe.throw(_("Payment Request already exist"))
+	
+	def validate_currency(self):
+		ref_doc = frappe.get_doc(self.reference_doctype, self.reference_name)
+		if ref_doc.currency != frappe.db.get_value("Account", self.payment_account, "account_currency"):
+			frappe.throw(_("Transaction currency is not simillar to Gateway Currency"))
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: Transaction currency must be same as Payment Gateway currency

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+		
+	def send_email(self):
+		"""send email with payment link"""
+		frappe.sendmail(recipients=self.email_to, sender=None, subject=self.subject,
+			message=self.get_message(), attachments=[frappe.attach_print(self.reference_doctype, 
+			self.reference_name, file_name=self.reference_name, print_format=self.print_format)])
+						
+	def get_message(self):
+		"""return message with payment gateway link"""
+		return  cstr(self.message) + """ <a href="%s"> Click here to pay </a>"""%self.payment_url
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: Make this configurable

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+	if args.submit_doc:
+		pr.insert(ignore_permissions=True)
+		pr.submit()
+		return pr
+	
+	return pr.as_dict()
+
+def get_reference_doc_details(dt, dn):
+	""" return reference doc Sales Order/Sales Invoice"""
+	return frappe.get_doc(dt, dn)
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: function not required

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+
+class PaymentRequest(Document):		
+	def validate(self):
+		self.validate_payment_request()
+		self.validate_currency()
+
+	def validate_payment_request(self):
+		if frappe.db.get_value("Payment Request", {"reference_name": self.reference_name, 
+			"name": ("!=", self.name), "status": ("not in", ["Initiated", "Paid"]), "docstatus": 1}, "name"):
+			frappe.throw(_("Payment Request already exist"))
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: Payment Request already exists for this {0}

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+	return frappe.db.get_value("Payment Gateway Account", {"is_default": 1}, 
+		["name", "gateway", "payment_account", "message"])
+
+@frappe.whitelist()
+def get_print_format_list(ref_doctype):
+	print_format_list = ["Standard"]
+	
+	print_format_list.extend(list(chain.from_iterable(frappe.db.sql("""select name from `tabPrint Format` 
+		where doc_type=%s""", ref_doctype, as_list=1))))
+
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: `[p.name for p in frappe.get_all("Print Format", filters={"doc_type": ref_doctype})]`

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 12]
```diff
@@ -0,0 +1,190 @@
...
+		pass
+	
+	def on_update_after_submit(self):
+		pass
+	
+	def set_status(self):
+		pass
+	
+	def send_payment_request(self):
+		self.payment_url = get_url("/api/method/erpnext.accounts.doctype.payment_request.payment_request.gererate_payemnt_request?name={0}".format(self.name))
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]:
```
def send_payment_request(self):
  if self.payment_gateway:
    self.payment_url = frappe.call_hook("get_payment_gateway_url")(self)
```

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 13]
```diff
@@ -0,0 +1,190 @@
...
+	
+	print_format_list.extend(list(chain.from_iterable(frappe.db.sql("""select name from `tabPrint Format` 
+		where doc_type=%s""", ref_doctype, as_list=1))))
+	
+	return {
+		"print_format": print_format_list
+	}
+	
+@frappe.whitelist(allow_guest=True)
+def gererate_payemnt_request(name):
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: fix spelling

File: erpnext/accounts/doctype/payment_request/payment_request.py [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+	doc = frappe.get_doc("Payment Request", name)
+	if doc.gateway == "PayPal":
```
Comment (rmehta) [inline erpnext/accounts/doctype/payment_request/payment_request.py]: payment_url = frappe.get_doc("Payment Gateway", self.payment_gateway).run_method("get_payment_url")

**Pull request closed without merging**